### PR TITLE
[WIP] feat: introduce and use electron-redux when possible

### PR DESCRIPTION
**Pull request by Shou**
We make use of the `electron-redux` library when possible, and otherwise
fall back to a plain instance of redux.

Changelog-Entry: Introduce and use electron-redux when possible.

### CODE REVIEW

**Comment by Shou** [comment]
All the redux-related tests are failing here, hence the WIP tag. Looks like either there is no main process in the tests, or asynchrony is being introduced and mucking things up. It works outside the tests.

File: lib/gui/app.js [hunk 1]
```diff
@@ -177,7 +177,7 @@ app.run(() => {
       `(eta ${currentFlashState.eta}s)`
     ].join(' '));
 
-    windowProgress.set(currentFlashState.percentage);
+    windowProgress.set(currentFlashState);
```
Comment (lurch) [inline lib/gui/app.js]: Hmm? Did this change accidentally slip through from #1438 ?

File: lib/shared/store.js [hunk 2]
```diff
@@ -486,7 +487,46 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
...
+// We are in the main Electron process
+} else if (process) {
+  console.log('MAIN REDUX');
+  const store = _.merge(redux.createStore(storeReducer, DEFAULT_STATE, redux.applyMiddleware(
+    electronRedux.triggerAlias,
+    electronRedux.forwardToRenderer
+  )), {
+    Actions: ACTIONS,
+    Defaults: DEFAULT_STATE
+  });
```
Comment (lurch) [inline lib/shared/store.js]: Looks like this `{ Actions: ACTIONS, Defaults: DEFAULT_STATE }` is repeated 3 separate times. I dunno what this code is doing, but would it be worth making a new constant for that?

**Comment by lurch** [comment]
Looks like the majority of this diff is just renaming a `data` attribute to `payload`. Apologies if this is a dumb question, but is that a necessary change?

**Comment by Shou** [comment]
`electron-redux` explains the change to `payload` as [FSA compliance](https://github.com/hardchor/electron-redux#actions), and indeed it doesn't work without the change.

File: lib/gui/app.js (hunk 1 above)
Comment (Shou) [inline lib/gui/app.js]: Well spotted

**Comment by Shou** [comment]
Shelving this for now.

File: lib/shared/store.js [hunk 3]
```diff
@@ -486,7 +487,42 @@ const storeReducer = (state = DEFAULT_STATE, action) => {
...
+};
+
+// We're not in Electron
+if (typeof process === 'undefined') {
+  console.log('PLAIN REDUX');
+  module.exports = _.merge(redux.createStore(storeReducer, DEFAULT_STATE), storeActionsAndDefaults);
+
+// We are in the renderer process of Electron
+} else if (_.get(process, [ 'type' ]) === 'renderer') {
+  console.log('RENDERER REDUX');
```
Comment (lurch) [inline lib/shared/store.js]: If this ever gets re-opened again, you'll need to remember to remove these `console.log`s ;-)

**Pull request closed without merging**